components/movementsensor/mpu6050/mpu6050.go

```diff
@@ -0,0 +1,292 @@
+	}
+	defer func() {
+		err := handle.Close()
+		if err != nil {
+			mpu.logger.Error(err)
+		}
+	}()
+
+	return handle.WriteByteData(ctx, register, value)
+}
+
+// A helper function: takes 2 bytes and reinterprets them as a big-endian signed integer.
+func toSignedValue(data []byte) int {
+	return int(int16(binary.BigEndian.Uint16(data)))
+}
+
+// Given a value, scales it so that the range of int16s becomes the range of +/- maxValue.
+func setScale(value int, maxValue float64) float64 {
+	return float64(value) * maxValue / (1 << 15)
+}
+
+// A helper function to abstract out shared code: takes 6 bytes and gives back AngularVelocity.
+func toAngularVelocity(data []byte) spatialmath.AngularVelocity {
+	gx := toSignedValue(data[0:2])
+	gy := toSignedValue(data[2:4])
+	gz := toSignedValue(data[4:6])
+
+	maxRotation := 250.0 // Degrees per second
+	return spatialmath.AngularVelocity{
+		X: setScale(gx, maxRotation),
+		Y: setScale(gy, maxRotation),
+		Z: setScale(gz, maxRotation),
+	}
+}
+
+// A helper function that takes 6 bytes and gives back linear acceleration.
+func toLinearAcceleration(data []byte) r3.Vector {
+	x := toSignedValue(data[0:2])
+	y := toSignedValue(data[2:4])
+	z := toSignedValue(data[4:6])
+
+	// The scale is +/- 2G's, but our units should be mm/sec/sec.
+	maxAcceleration := 2.0 * 9.81 /* m/sec/sec */ * 1000.0 /* mm/m */
+	return r3.Vector{
+		X: setScale(x, maxAcceleration),
+		Y: setScale(y, maxAcceleration),
+		Z: setScale(z, maxAcceleration),
+	}
+}
+
+func (mpu *mpu6050) AngularVelocity(ctx context.Context, extra map[string]interface{}) (spatialmath.AngularVelocity, error) {
+	mpu.mu.Lock()
+	defer mpu.mu.Unlock()
+
+	// The gyroscope measurements are in registers 67 to 72 (2 bytes for each axis).
+	rawData, err := mpu.readBlock(ctx, 67, 6)
+	if err != nil {
+		return spatialmath.AngularVelocity{}, err
+	}
+	return toAngularVelocity(rawData), nil
+}
+
+func (mpu *mpu6050) LinearVelocity(ctx context.Context, extra map[string]interface{}) (r3.Vector, error) {
+	return r3.Vector{}, movementsensor.ErrMethodUnimplementedLinearVelocity
+}
+
+func (mpu *mpu6050) Orientation(ctx context.Context, extra map[string]interface{}) (spatialmath.Orientation, error) {
+	return nil, movementsensor.ErrMethodUnimplementedOrientation
+}
+
+func (mpu *mpu6050) CompassHeading(ctx context.Context, extra map[string]interface{}) (float64, error) {
+	return 0, movementsensor.ErrMethodUnimplementedCompassHeading
+}
+
+func (mpu *mpu6050) Position(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+	return geo.NewPoint(0, 0), 0, movementsensor.ErrMethodUnimplementedPosition
+}
+
+func (mpu *mpu6050) Accuracy(ctx context.Context, extra map[string]interface{}) (map[string]float32, error) {
+	return map[string]float32{}, movementsensor.ErrMethodUnimplementedAccuracy
+}
+
+func (mpu *mpu6050) Readings(ctx context.Context, extra map[string]interface{}) (map[string]interface{}, error) {
+	mpu.mu.Lock()
+	defer mpu.mu.Lock()
```

Comment:
:facepalm: Well _there's_ yer problem! After calling this function, the mutex never gets unlocked, which means no other values can be read afterwards.

Reply:
...and it's worse than that! By deferring a second Lock(), we have a deadlock in the deferred part of the function, which is why we don't even get a response from the first call to the readings.